segfault in `Rust::AST::RangePatternBound::clone_range_pattern_bound` range pattern in macro

<!--
File: /tmp/im/7EB53A3B29E8DA67C7299D222B23E356F2F9CB24EF69C89F6A03DACD4E079366.rs
-->

auto-reduced (treereduce-rust):
````rust
//@compile-flags: -frust-incomplete-and-experimental-compiler-do-not-use
macro_rules! m {
    () => {
        c..
    };
}

fn main() {
    match [1, 1, 2] {
        [m!()] => {}
    }
}

````

original:
````rust
#![feature()]

macro_rules! m {
    () => ( c.. );
}

fn main() {
    match [1, 1, 2] {
        [ c1 ] => {}
        [ m!() ] => {}
    }
}
````

Version information:

4db0cf74f64086e2bb2eff029c597239df8442e1



Command:
`/home/matthias/vcs/github/gccrs/gccrs-build/gcc/crab1 -frust-incomplete-and-experimental-compiler-do-not-use`

<details><summary><strong>Program output</strong></summary>
<p>

```
/tmp/icemaker_global_tempdir.J4KdGj8QZ6tS/rustc_testrunner_tmpdir_reporting.ZBHIfJ49wxjP/mvce.rs:4:5: error: token type ‘}’ cannot be parsed as range pattern bound
    4 |     };
      |     ^
crab1: internal compiler error: Segmentation fault
0x301e249 internal_error(char const*, ...)
	../../gcc/diagnostic-global-context.cc:786
0x199e5df crash_signal
	../../gcc/toplev.cc:323
0xe1dbcf Rust::AST::RangePatternBound::clone_range_pattern_bound() const
	../../gcc/rust/ast/rust-pattern.h:248
0xe1dbcf Rust::AST::RangePattern::RangePattern(Rust::AST::RangePattern const&)
	../../gcc/rust/ast/rust-pattern.h:413
0xe1dbcf Rust::AST::RangePattern::clone_pattern_impl() const
	../../gcc/rust/ast/rust-pattern.h:471
0xde210c Rust::AST::Pattern::clone_pattern() const
	../../gcc/rust/ast/rust-ast.h:1427
0xde210c Rust::AST::SingleASTNode::SingleASTNode(Rust::AST::SingleASTNode const&)
	../../gcc/rust/ast/rust-ast.cc:76
0xea6845 void std::_Construct<Rust::AST::SingleASTNode, Rust::AST::SingleASTNode const&>(Rust::AST::SingleASTNode*, Rust::AST::SingleASTNode const&)
	/usr/include/c++/15.2.1/bits/stl_construct.h:133
0xea6845 Rust::AST::SingleASTNode* std::__do_uninit_copy<Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode*>(Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode*)
	/usr/include/c++/15.2.1/bits/stl_uninitialized.h:145
0xea6845 Rust::AST::SingleASTNode* std::uninitialized_copy<Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode*>(Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode*)
	/usr/include/c++/15.2.1/bits/stl_uninitialized.h:266
0xea6845 Rust::AST::SingleASTNode* std::__uninitialized_copy_a<Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode*, Rust::AST::SingleASTNode>(Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode*, std::allocator<Rust::AST::SingleASTNode>&)
	/usr/include/c++/15.2.1/bits/stl_uninitialized.h:637
0xea6845 void std::vector<Rust::AST::SingleASTNode, std::allocator<Rust::AST::SingleASTNode> >::_M_range_initialize_n<Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode const*>(Rust::AST::SingleASTNode const*, Rust::AST::SingleASTNode const*, unsigned long)
	/usr/include/c++/15.2.1/bits/stl_vector.h:1989
0xea6845 std::vector<Rust::AST::SingleASTNode, std::allocator<Rust::AST::SingleASTNode> >::vector(std::initializer_list<Rust::AST::SingleASTNode>, std::allocator<Rust::AST::SingleASTNode> const&)
	/usr/include/c++/15.2.1/bits/stl_vector.h:712
0xea6845 transcribe_pattern
	../../gcc/rust/expand/rust-macro-expand.cc:1020
0xea6845 transcribe_context
	../../gcc/rust/expand/rust-macro-expand.cc:1065
0xea6845 Rust::MacroExpander::transcribe_rule(Rust::AST::MacroRulesDefinition&, Rust::AST::MacroRule&, Rust::AST::DelimTokenTree&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, Rust::MatchedFragmentContainer*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, Rust::MatchedFragmentContainer*> > >&, Rust::AST::InvocKind, Rust::MacroExpander::ContextType)
	../../gcc/rust/expand/rust-macro-expand.cc:1151
0xea94c4 Rust::MacroExpander::expand_decl_macro(unsigned long, Rust::AST::MacroInvocData&, Rust::AST::MacroRulesDefinition&, Rust::AST::InvocKind)
	../../gcc/rust/expand/rust-macro-expand.cc:127
0xea9968 Rust::MacroExpander::expand_invoc(Rust::AST::MacroInvocation&, Rust::AST::InvocKind)
	../../gcc/rust/expand/rust-macro-expand.cc:329
0xeb6b5f Rust::ExpandVisitor::maybe_expand_pattern(std::unique_ptr<Rust::AST::Pattern, std::default_delete<Rust::AST::Pattern> >&)
	../../gcc/rust/expand/rust-expand-visitor.cc:391
0xeb6e4e Rust::ExpandVisitor::visit(Rust::AST::SlicePatternItemsNoRest&)
	../../gcc/rust/expand/rust-expand-visitor.cc:1050
Please submit a full bug report, with preprocessed source (by using -freport-bug).
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.

```

</p>
</details>

--- a/gcc/testsuite/rust/compile/issue-4159.rs
+++ b/gcc/testsuite/rust/compile/issue-4159.rs
@@ -1,0 +1,12 @@
+#![feature(no_core)]
+#![no_core]
+#![feature()]
+macro_rules! m {
+    () => ( c.. ); // { dg-error "token type .* cannot be parsed as range pattern bound" }
+}
+fn main() {
+    match [1, 1, 2] {
+        [ c1 ] => {}
+        [ m!() ] => {}
+    }
+}

